include/psa/crypto_compat.h

```diff
@@ -269,6 +291,234 @@ MBEDTLS_PSA_DEPRECATED static inline psa_status_t psa_asymmetric_verify( psa_key
 #define PSA_ALG_AEAD_WITH_TAG_LENGTH( aead_alg, tag_length ) \
     MBEDTLS_DEPRECATED_CONSTANT( psa_algorithm_t, PSA_ALG_AEAD_WITH_SHORTENED_TAG( aead_alg, tag_length ) )
 
+/*
+ * Deprecated PSA AEAD output size macros (PSA Crypto API  <= 1.0 beta3)
+ */
+
+/** The tag size for an AEAD algorithm, in bytes.
+ *
+ * \param alg                 An AEAD algorithm
+ *                            (\c PSA_ALG_XXX value such that
+ *                            #PSA_ALG_IS_AEAD(\p alg) is true).
+ *
+ * \return                    The tag size for the specified algorithm.
+ *                            If the AEAD algorithm does not have an identified
+ *                            tag that can be distinguished from the rest of
+ *                            the ciphertext, return 0.
+ *                            If the AEAD algorithm is not recognized, return 0.
+ */
+#define PSA_AEAD_TAG_LENGTH_1_ARG( alg )                                         \
+    MBEDTLS_DEPRECATED_CONSTANT( size_t,                                         \
+        PSA_ALG_IS_AEAD( alg ) ?                                                 \
+        ( (alg) & PSA_ALG_AEAD_TAG_LENGTH_MASK ) >> PSA_AEAD_TAG_LENGTH_OFFSET : \
+        0 )
+
+#if defined(PSA_AEAD_TAG_LENGTH)
+#undef PSA_AEAD_TAG_LENGTH
+#endif
+
+#define PSA_AEAD_TAG_LENGTH( ... ) \
+    MBEDTLS_EXPAND( \
+        MBEDTLS_STATIC_ASSERT_THEN_RETURN( \
```

Comment:
I don't fully understand what's going on here. I think the point of the static assert is to cause an error if the macro is called with a number of arguments other than 1 or 3, but I don't fully understand how this works, and the error message in that case is incomprehensible.

Can't we do something simpler? Something like
```
#define PSA_AEAD_TAG_LENGTH_ARITY_CASE( arg1, arg2, arg3, arg_count, ... ) \
    ( arg_count == 1 ? PSA_AEAD_TAG_LENGTH_1( arg1 ) :                  \
      arg_count == 3 ? PSA_AEAD_TAG_LENGTH_3( arg1, arg2, arg3 ) :      \
      PSA_AEAD_TAG_LENGTH_called_with_0_or_2_arguments )
#define PSA_AEAD_TAG_LENGTH( ... )                                      \
    MBEDTLS_EXPAND( MBEDTLS_ARGS_LEN( __VA_ARGS__ ) <= 3 ?              \
                    PSA_AEAD_TAG_LENGTH_ARITY_CASE( __VA_ARGS__, 3, 2, 1 ) : \
                    PSA_AEAD_TAG_LENGTH_called_with_too_many_arguments )
```

Reply:
The main problem I see with your solution is `PSA_AEAD_TAG_LENGTH_called_with_too_many_arguments`.

I wanted to (reliably) cause a compilation error if someone used too many arguments. In your example, this problem would only be detectable at runtime - if `PSA_AEAD_TAG_LENGTH_called_with_too_many_arguments` is valid C code, it will compile without problem, and if it is not, it will error out even if the branch is not taken.

As for the rest, the way I've set it up the inappropriate macro never gets expanded. This reduces the total expanded size of the macro, and prevents stuff like `PSA_AEAD_TAG_LENGTH_3(foo, 3, 2)` being compiled in the untaken branch. 

I don't think this would cause any issues in our case, since these macros are only dealing with integers, but it would blow up if one of the macros was supposed to take something like a pointer or an array as an argument.

Reply:
You would also need to use `PSA_AEAD_TAG_LENGTH_ARITY_CASE( __VA_ARGS__, 3, 2, 1, 0 )` (padding with a 0), since an empty `...` argument is actually a gcc extension - ISO C does not allow it.

Basically, my code uses equivalent logic to your ARITY_CASE, just that I use PSA_AEAD_TAG_LENGTH_1_ARG and `PSA_AEAD_TAG_LENGTH_3_ARG` as the "counter" values instead of 3, 2, 1 and 0. The only values that are actually called are the first  `PSA_AEAD_TAG_LENGTH_3_ARG` (equivalent to 3 in your code), and the first PSA_AEAD_TAG_LENGTH_1_ARG (equivalent to 1 in your code), the other two are just the padding required to make this work, with values chosen to produce informative error messages if someone calls the macro with 2 arguments.

Reply:
Yes, I see, thank you, what I attempted to do (leverage C expressions to detect errors) is fundamentally impossible.

I think that this use of the preprocessor is [too clever to be maintainable](https://www.quotes.net/quote/57502). I don't think it's worth the effort. I know you've already invested a lot of effort in making it work, but there will be maintenance cost: risk of encountering compiler bugs, time taken in debugging expressions involving these macros, supporting users who don't understand error messages...

I'm not going to veto it, but I will not spend the time to review it properly, which would require fully understanding it. As a gatekeeper, if two reviewers assert that they fully understand it and that they don't mind if [a future maintainer of the library who is a psychopath and knows their home address](https://github.com/ARMmbed/mbedtls-docs/blob/main/reviews/How%20to%20be%20an%20effective%20Mbed%20TLS%20reviewer.pdf) has to work on these macros, I'm ok with merging it.

As an architect, I recommend an alternative approach: keep the old macro as a deprecated `PSA_AEAD_TAG_LENGTH_1`, introduce the new macro as `PSA_AEAD_TAG_LENGTH` (and similarly for the other macros whose usage is changing). Document this in the changelog as an API change, listing the macros that need to be adapted. We do have a few precedents of doing not-fully-compatible API changes to align with the PSA specification.

Reply:
I agree with the "As a architect" proposal. That's a much better way to go forward to me and the "keep the old macro as a deprecated PSA_AEAD_TAG_LENGTH_1" proposal mitigates significantly the negative impact on users: their code will not compile and they just have to add a "_1" to the modified macros to make it back to compile and work as before.

Otherwise, I don't feel I can be one of the reviewers that "don't mind if a future maintainer of the library who is a psychopath and knows their home address has to work on these macros".